# Homework 3
## ECE204 Data Science & Engineering

---
## Question 1
> A number N is _divisible_ by P if P divides N evenly. For example, 15 is divisible by 1, 3, 5, and 15. The numbers [1,3,5,15] are called the "divisors" of 15. Your task is to create a function that returns the sum of all the divisors of a number. For example, if you call the function "F", you should find that F(15) returns 24.
> We can figure out if N is divisible by P by using Python's "modulo" function. The syntax for this is:
> ```python
> N % P == 0    # this returns True if P divides N
>```
> <br>For more information about the modulo (%) function in Python, see [this link](https://www.pythoncentral.io/using-python-to-check-for-odd-or-even-numbers/).
>
> Write the function "F" and test it to make sure that `F(15) == 24`. Call the function with the argument given to you in Canvas.

In [2]:
def F(N):
    divisor_sum = 0
    for i in range(1, N + 1):
        if N % i == 0:
            divisor_sum += i
    return divisor_sum

# Test the function with F(15) and F(18045)
result_15 = F(15)
result_18045 = F(18045)

print("F(15) =", result_15)
print("F(18045) =", result_18045)


F(15) = 24
F(18045) = 31356


---
## Question 2
> List comprehensions we saw in class looked like this:
> ```python
> [ (do this) for (item in collection) ]
> ```
> <br>But what if we only want to select the items that satisfy an additional condition as well? We can use the syntax:
> ```python
> [ (do this) for (item in collection) if (true/false condition) ]
> ```
> <br>**Use this technique to find the sum of numbers divisible by 7 or 11 and report this result. Be sure to use the range of vlaues (inclusive) given to you in Canvas**. You may want to start with a smaller number first to make sure your list comprehension is doing the right thing!
>
> `NOTE:` If a number is divisible by both 7 and 11, add it only **once**.
>
> Hint: The answer for numbers 1 to 30 would be: 7 + 11 + 14 + 21 + 22 + 28 = 103

In [3]:
# your code here
result = sum(num for num in range(196, 10161) if num % 7 == 0 or num % 11 == 0)
print("Sum of numbers divisible by 7 or 11 from 196 to 10160:", result)


Sum of numbers divisible by 7 or 11 from 196 to 10160: 11394828


---
## Question 3
> The square of the sum of the first ten natural numbers is
>
> $$(1 + 2 + ... + 10)^2 = 55^2 = 3025$$
> 
>   The sum of the squares of the first ten natural numbers is
>
> $$1^2 + 2^2 + ... + 10^2 = 385$$
>
> Hence the difference between the square of the sum of first ten natural numbers and the sum of the squares is 3025 − 385 = 2640.
>
> **Find the difference between the square of the sum of the first X natural numbers and the sum of the squares. Note that X is the value given to you in Canvas**


In [4]:
# your code here
n = 165

# Calculate the square of the sum
square_of_sum = sum(range(1, n + 1))**2

# Calculate the sum of the squares
sum_of_squares = sum(i**2 for i in range(1, n + 1))

# Calculate the difference
difference = square_of_sum - sum_of_squares

print("The difference between the square of the sum and the sum of squares for the first", n, "natural numbers is:", difference)


The difference between the square of the sum and the sum of squares for the first 165 natural numbers is: 186042010


---
## Question 4
> Out of all the points in the list of `candidates`, find the one that is the furthest (in Euclidean distance) from the `given` point? Note: the `given` point is provided for you in Canvas. 
> What is this furthest / highest distance? **Report this furthest distance rounded to two numbers after the decimal.**
>
> `Hint: np.linalg.norm( )`

In [5]:
candidates = [
    (0.0, 1.0),
    (0.32, 0.95),
    (0.61, 0.79),
    (0.84, 0.55),
    (0.97, 0.25),
    (1.0, -0.08),
    (0.92, -0.4),
    (0.74, -0.68),
    (0.48, -0.88),
    (0.16, -0.99),
    (-0.16, -0.99),
    (-0.48, -0.88),
    (-0.74, -0.68),
    (-0.92, -0.4),
    (-1.0, -0.08),
    (-0.97, 0.25),
    (-0.84, 0.55),
    (-0.61, 0.79),
    (-0.32, 0.95),
    (-0.0, 1.0)
]

In [8]:
import numpy as np

# Given point
given_point = [3, 1]

# List of candidate points
candidates = [
    (0.0, 1.0),
    (0.32, 0.95),
    (0.61, 0.79),
    (0.84, 0.55),
    (0.97, 0.25),
    (1.0, -0.08),
    (0.92, -0.4),
    (0.74, -0.68),
    (0.48, -0.88),
    (0.16, -0.99),
    (-0.16, -0.99),
    (-0.48, -0.88),
    (-0.74, -0.68),
    (-0.92, -0.4),
    (-1.0, -0.08),
    (-0.97, 0.25),
    (-0.84, 0.55),
    (-0.61, 0.79),
    (-0.32, 0.95),
    (-0.0, 1.0)
]

# Calculate the Euclidean distance for each candidate point
distances = [np.linalg.norm(np.array(candidate) - np.array(given_point)) for candidate in candidates]

# Find the maximum distance
max_distance = max(distances)

# Print the maximum distance rounded to two decimal places
print("The furthest distance from the given point is:", round(max_distance, 2))


The furthest distance from the given point is: 4.16


---
## Question 5
> `celsius.csv` contains a list of temperatures in Celsius. **What is the mean Fahrenheit temperature? (Report your answer rounded to two numbers after the decimal)**
>
> The Celsuis to Fahrenheit conversion is given by `fahrenheit = 1.8 * celsius + 32`.

In [10]:
# given code
import pandas as pd
# Import the data file and have a look!
s = pd.read_csv("celsius.csv")
s.head()

temps
0  54.88
1  71.52
2  60.28
3  54.49
4  42.37

In [12]:
import csv

# Initialize an empty list to store Celsius temperatures
celsius_temperatures = []

# Read Celsius temperatures from the CSV file
with open('celsius.csv', 'r') as file:
    csv_reader = csv.reader(file)
    next(csv_reader)  # Skip the header row if it exists
    for row in csv_reader:
        celsius_temperatures.append(float(row[0]))

# Convert Celsius to Fahrenheit and calculate the mean
fahrenheit_temperatures = [1.8 * c + 32 for c in celsius_temperatures]
mean_fahrenheit = sum(fahrenheit_temperatures) / len(fahrenheit_temperatures)

# Print the mean Fahrenheit temperature rounded to two decimal places
print("The mean Fahrenheit temperature is:", round(mean_fahrenheit, 2))


The mean Fahrenheit temperature is: 121.36


---
## Question 6
> **What is the fastest method to convert from Celsius to Fahrenheit?**
>
> * a list comphrension used to make a list of all the converted temperatures
> * using `Series.apply` to convert each element of the Series
> * directly evaluating (`1.8*T + 32`), where `T` is the Series of temperatures
>
> `NOTE:` Use `%timeit` to do your timing analysis. Use the `celsius.csv` data file from the previous problem.

In [18]:
import pandas as pd

# Load data from the CSV file
df = pd.read_csv('celsius.csv')

# Convert Celsius to Fahrenheit using direct evaluation
df['Fahrenheit'] = 1.8 * df['temps'] + 32

# Print the DataFrame with Celsius and Fahrenheit temperatures
print(df)


      temps  Fahrenheit
0     54.88     130.784
1     71.52     160.736
2     60.28     140.504
3     54.49     130.082
4     42.37     108.266
...     ...         ...
9995  55.04     131.072
9996  39.72     103.496
9997  75.84     168.512
9998   2.38      36.284
9999  81.36     178.448

[10000 rows x 2 columns]


---
## Question 7
> **What is one of the centers in `features.csv` when using Scikit-Learn's KMeans?** They are listed as `[feature0, feature1]`, and `KMeans` should be used as `KMeans(n_clusters=4, random_state=42)`. These numbers are rounded -- pick the closest one in the list.
>
> Hint: read the documentation at https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
>
> * [170  67]
> * [170  88]
> * [201  88]
> * [201  67]
> * [-131  748]
> * [-131  728]
> * [-161  728]
> * [-161  748]

In [14]:
# Read the dataset and take a look!
df = pd.read_csv("features.csv")
df.head()

feature0  feature1
0    124.30    172.35
1   -135.85    755.16
2    109.56    483.43
3   -109.52    782.18
4    153.27    156.09

In [15]:
# your code here
import pandas as pd
from sklearn.cluster import KMeans


# Initialize KMeans with n_clusters=4 and random_state=42
kmeans = KMeans(n_clusters=4, random_state=42)

# Fit the KMeans model
kmeans.fit(df)

# Get one of the cluster centers
center = kmeans.cluster_centers_[0]  # You can change the index to get different centers

print("One of the cluster centers is:", center)


C:\Users\16089\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


One of the cluster centers is: [200.6812      87.76973333]


---
## Question 8
> **Using the same setup as before, which cluster center is the point `[-100, 850]` nearest to?** <br>
>
> `NOTE:` **choose a valid cluster center**
>
> * [-157,  286]
> * [-91,  743]
> * [ 201,   87]
> * [  -133,  778]

In [16]:
# your code here
from sklearn.cluster import KMeans

# Point to be assigned to the nearest cluster
point = [-100, 850]

# Initialize KMeans with n_clusters=4 and random_state=42
kmeans = KMeans(n_clusters=4, random_state=42)

# Fit the KMeans model
kmeans.fit(df) 

# Use the predict method to find the nearest cluster
nearest_cluster = kmeans.predict([point])[0]

# List of cluster centers
cluster_centers = kmeans.cluster_centers_

# Find the cluster center that the point is nearest to
nearest_center = cluster_centers[nearest_cluster]

print("The point [-100, 850] is nearest to cluster center:", nearest_center)


The point [-100, 850] is nearest to cluster center: [-133.15813333  778.09466667]


C:\Users\16089\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
C:\Users\16089\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:465: UserWarning: X does not have valid feature names, but KMeans was fitted with feature names
  warnings.warn(


---